/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import strmv = require( './index' );


// TESTS //

// The function returns a Float32Array...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectType Float32Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv( 10, 'upper', 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( true, 'upper', 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( false, 'upper', 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( null, 'upper', 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( undefined, 'upper', 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( [], 'upper', 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( {}, 'upper', 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( ( x: number ): number => x, 'upper', 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a string...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv( 'row-major', 10, 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', true, 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', false, 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', null, 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', undefined, 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', [], 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', {}, 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', ( x: number ): number => x, 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a string...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv( 'row-major', 'upper', 10, 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', true, 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', false, 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', null, 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', undefined, 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', [], 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', {}, 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', ( x: number ): number => x, 'unit', 10, A, 10, x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a string...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv( 'row-major', 'upper', 'none', 10, 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', true, 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', false, 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', null, 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', undefined, 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', [], 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', {}, 10, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', ( x: number ): number => x, 10, A, 10, x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv( 'row-major', 'upper', 'none', 'unit', '10', A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', true, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', false, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', null, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', undefined, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', [], A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', {}, A, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', ( x: number ): number => x, A, 10, x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a Float32Array...
{
	const x = new Float32Array( 10 );

	strmv( 'row-major', 'upper', 'none', 'unit', 10, 10, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, '10', 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, true, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, false, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, null, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, undefined, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, [ '1' ], 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, {}, 10, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, ( x: number ): number => x, 10, x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a number...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, '10', x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, true, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, false, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, null, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, undefined, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, [], x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, {}, x, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, ( x: number ): number => x, x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a eighth argument which is not a Float32Array...
{
	const A = new Float32Array( 20 );

	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, 10, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, '10', 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, true, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, false, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, null, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, undefined, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, [ '1' ], 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, {}, 1 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, ( x: number ): number => x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, '10' ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, true ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, false ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, null ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, undefined ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, [] ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, {} ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv(); // $ExpectError
	strmv( 'row-major' ); // $ExpectError
	strmv( 'row-major', 'upper' ); // $ExpectError
	strmv( 'row-major', 'upper', 'none' ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit' ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10 ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x ); // $ExpectError
	strmv( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, 1, 1 ); // $ExpectError
}

// Attached to main export is an `ndarray` method which returns a Float32Array...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, 1, 0 ); // $ExpectType Float32Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv.ndarray( 10, 'upper', 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( true, 'upper', 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( false, 'upper', 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( null, 'upper', 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( undefined, 'upper', 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( [], 'upper', 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( {}, 'upper', 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( ( x: number ): number => x, 'upper', 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a string...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv.ndarray( 'row-major', 10, 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', true, 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', false, 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', null, 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', undefined, 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', [], 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', {}, 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', ( x: number ): number => x, 'none', 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a string...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv.ndarray( 'row-major', 'upper', 10, 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', true, 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', false, 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', null, 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', undefined, 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', [], 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', {}, 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', ( x: number ): number => x, 'unit', 10, A, 10, x, 1 , 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a string...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv.ndarray( 'row-major', 'upper', 'none', 10, 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', true, 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', false, 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', null, 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', undefined, 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', [], 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', {}, 10, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', ( x: number ): number => x, 10, A, 10, x, 1 , 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', '10', A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', true, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', false, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', null, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', undefined, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', [], A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', {}, A, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', ( x: number ): number => x, A, 10, x, 1 , 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a Float32Array...
{
	const x = new Float32Array( 10 );

	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, 10, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, '10', 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, true, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, false, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, null, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, undefined, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, [ '1' ], 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, {}, 10, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, ( x: number ): number => x, 10, x, 1 , 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a number...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, '10', x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, true, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, false, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, null, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, undefined, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, [], x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, {}, x, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, ( x: number ): number => x, x, 1 , 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a eighth argument which is not a Float32Array...
{
	const A = new Float32Array( 20 );

	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, 10, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, '10', 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, true, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, false, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, null, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, undefined, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, [ '1' ], 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, {}, 1 , 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, ( x: number ): number => x, 1 , 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, '10', 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, true, 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, false, 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, null, 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, undefined, 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, [], 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, {}, 0 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, ( x: number ): number => x, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a tenth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, 1, '10' ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, 1, true ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, 1, false ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, 1, null ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, 1, undefined ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, 1, [] ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, 1, {} ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, 1, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	strmv.ndarray(); // $ExpectError
	strmv.ndarray( 'row-major' ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper' ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none' ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit' ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, 1 ); // $ExpectError
	strmv.ndarray( 'row-major', 'upper', 'none', 'unit', 10, A, 10, x, 1, 0, 10 ); // $ExpectError
}
